src/nxt_router.c

```diff
@@ -1602,6 +1599,21 @@ nxt_router_conf_create(nxt_task_t *task, nxt_router_temp_conf_t *tmcf,
         rtcf->threads = nxt_ncpu;
     }
 
+    log = nxt_conf_get_path(root, &log_path);
+    if (log != NULL) {
+        nxt_memzero(&lgcf, sizeof(lgcf));
+        ret = nxt_conf_map_object(mp, log,
+                                  nxt_router_log_conf,
+                                  nxt_nitems(nxt_router_log_conf),
+                                  &lgcf);
+        if (ret != NXT_OK) {
+            nxt_alert(task, "log map error");
+            return NXT_ERROR;
+        }
+
+        task->log->level = lgcf.level;
```

Comment:
Here, I checked that `task->log` is really a pointer to `nxt_main_log`, so we're really changing the main log level that affects all of Unit.  I did some printf `%p` with that pointer, and also `&nxt_main_log`, and both printed the same value.

Reply:
Every process has its own main log, correct?

Reply:
Well, I guess they do (unless we do some shared memory magic for it).  So, this should affect the one in the router, and all descendants of it.  Existing childs of the router will also probably keep the old log level.  Unless we use shared memory for the log structure, I don't think we can improve this further, right?

Reply:
What is the definite scope of the log level? We have main/controller/router/app processes.

Will it apply to all the processes or just some of them?

Reply:
I think I'd like it to apply to all processes.  However, I'm not sure what the current patches do.  I know if affects the entire router.  I don't know what happens with other processes.

Reply:
I suggest we test it before reviewing it.

Reply:
Now that I checked, the apps don't have log level conditionals.  They always print the logs.  So we can't modify them.

Reply:
> Maybe it's enough to modify the router for now.

We need to make it clear before reviewing it, right?

Reply:
Yes, so let's say that we only want to affect the router.

Reply:
We should update the confluence page:

Applications use `nxt_unit_log()`, which doesn't have a task, and doesn't check the log level of the message before printing it.

Reply:
```c
./nxt_unit.c:6578:
void
nxt_unit_log(nxt_unit_ctx_t *ctx, int level, const char *fmt, ...)
{
    int              log_fd, n;
    char             msg[NXT_MAX_ERROR_STR], *p, *end;
    pid_t            pid;
    va_list          ap;
    nxt_unit_impl_t  *lib;

    if (nxt_fast_path(ctx != NULL)) {
        lib = nxt_container_of(ctx->unit, nxt_unit_impl_t, unit);

        pid = lib->pid;
        log_fd = lib->log_fd;

    } else {
        pid = nxt_unit_pid;
        log_fd = STDERR_FILENO;
    }

    p = msg;
    end = p + sizeof(msg) - 1;

    p = nxt_unit_snprint_prefix(p, end, pid, level);

    va_start(ap, fmt);
    p += vsnprintf(p, end - p, fmt, ap);
    va_end(ap);

    if (nxt_slow_path(p > end)) {
        memcpy(end - 5, "[...]", 5);
        p = end;
    }

    *p++ = '\n';

    n = write(log_fd, msg, p - msg);
    if (nxt_slow_path(n < 0)) {
        fprintf(stderr, "Failed to write log: %.*s", (int) (p - msg), msg);
    }
}
```

The level is only used for the log prefix, but it's not used elsewhere.

Reply:
It's about the scope of the log level. Other places like the main process and controller also do possible logging.

Reply:
Based on the above information. Let's make a definitive scope of the log level first before reviewing patches.

>  1. I think I'd like it to apply to all processes. 
>  2. let's say that we only want to affect the router.

Which one do we choose?

Reply:
If we separate the router process from all processes. They could be different requirements and introduces different implementations. I mean if want to modify the main and controller processes in the future, we might rewrite it, right?

If we only affect the router processes, do we need to introduce a new log level? It seems we can reuse the existing log level since there is no log in the http route.

Anyway, in my feelings, the different definitions will determine different implementations.

Reply:
> If we separate the router process from all processes. They could be different requirements and introduces different implementations.

Yes, that's what I'm suggesting.  Implement them separately, with different requirements.

> I mean if want to modify the main and controller processes in the future, we might rewrite it, right?

Yes, we might rewrite it in the future, for implementing the other features, of course.

> 
> If we only affect the router processes, do we need to introduce a new log level? It seems we can reuse the existing log level since there is no log in the http route.

We need the new log level, because the default is `[info]`, which is already in use.  We don't want to print lines in `[info]` for every request. 

> 
> Anyway, in my feelings, the different definitions will determine different implementations.

Sure.

Reply:
If you only apply the patches for http route logging, it's not the right time to introduce a new log level since there is no logging in the http route yet.

BTW, is it good to have two different requirements for the logging?
I would suggest making them clear, it's a bit confusing for reviewers.

Reply:
> If you only apply the patches for http route logging, it's not the right time to introduce a new log level since there is no logging in the http route yet.

I don't understand you.  What do you mean by http route?  I see at least the following logging:

```c
$ grep -rn NXT_LOG_INFO src/nxt_router.c
323:    nxt_log(task, NXT_LOG_INFO, "router started");
```

> 
> BTW, is it good to have two different requirements for the logging? I would suggest making them clear, it's a bit confusing for reviewers.

Reply:
I mean the `nxt_http_route.c` file as the commit log "Log decisions about matching".
There is no logging in the matching before introducing the patches.

Reply:
I think it's better to make it clear what requirements we want to handle, then we can start reviewing.
And it makes sense for the current requirement that log level is for affecting all logs. But I don't know what it solves for the new requirement of only working for the router process, it looks like a newly introduced requirement and no one discuss it if it makes sense before.

Reply:
I wonder if you only add logging in the "http route/matching"? If not, what others did you introduce?

Reply:
> I wonder if you only add logging in the "http route/matching"? If not, what others did you introduce?

There are all these commits:

```
6ee17d9a Configuration: added "setting/log".
1d83f2b0 Router: factored out listeners creation loop.
aecc848e Router: factored out application creation loop.
1583fdb0 Router: logging request line.
3d564018 Router: logging the kind of element tried for a match.
e3d269f4 Router: refactored code.
77cffddb Router: logging when a "share" takes the "fallback" action.
e672703d Router: logging when control is passed to a "share".
9815e2a5 Router: expanded location abbreviation in log message.
330291cd Router: printing the application name in the debug log.
1d0e9951 Router: logging route array elements being evaluated.
9d811a8b Router: logging when control is passed to a route.
fb621ec4 Router: logging route whose action is selected.
f5fc24f3 Router: logging pattern matching decissions.
62564a14 Router: refactored code.
3dd8810b PCRE: moved structure definition to header.
f2d12359 Router: rewrote while loop as for loop.
296559df Log: added "diag" log level.
```

From those, the ones that add new logging lines are:

```
1583fdb0 Router: logging request line.
3d564018 Router: logging the kind of element tried for a match.
77cffddb Router: logging when a "share" takes the "fallback" action.
e672703d Router: logging when control is passed to a "share".
1d0e9951 Router: logging route array elements being evaluated.
9d811a8b Router: logging when control is passed to a route.
fb621ec4 Router: logging route whose action is selected.
f5fc24f3 Router: logging pattern matching decissions.
```

They affect the following files and functions:

```
src/nxt_h1proto.c:
    nxt_h1p_header_process()
src/nxt_http_route.c:
    nxt_http_route_match()
    nxt_http_route_rule()
    nxt_http_route_header()
    nxt_http_route_test_argument()
    nxt_http_route_scheme()
    nxt_http_route_test_cookie()
    nxt_http_route_handler()
    nxt_http_route_pattern()
src/nxt_http_static.c:
    nxt_http_static()
    nxt_http_static_next()
```

Reply:
So we changed three files, right?
1. src/nxt_h1proto.c.
```
nxt_log(task, NXT_LOG_DIAG, "http request line: \"%V %V %V\"", r->method,
            &r->target, &r->version);
```
It's not required or maybe it can be just a debugging.
2. src/nxt_http_static.c.
It adds debug information without logging.

3. nxt_http_route.c
Adding debug and logging.

As I mentioned, there is no logging in the nxt_http_route.c before. I prefer not to over-engineering if we only add logging in the nxt_http_route.c, I think the existing log level is enough.

Let the team make the decision of what we should resolve for users, good to discuss.

Reply:
> So we changed three files, right?
> 
>     1. src/nxt_h1proto.c.
> 
> 
> ```
> nxt_log(task, NXT_LOG_DIAG, "http request line: \"%V %V %V\"", r->method,
>             &r->target, &r->version);
> ```
> 
> It's not required or maybe it can be just a debugging.

No.  We do want that line.  In fact, that's probably one of the most important.  Otherwise, it's difficult to understand the rest of the logging.

> 
>     2. src/nxt_http_static.c.
>        It adds debug information without logging.
> 
>     3. nxt_http_route.c
>        Adding debug and logging.

Correct.

> 
> 
> As I mentioned, there is no logging in the nxt_http_route.c before. I prefer not to over-engineering if we only add logging in the nxt_http_route.c, I think the existing log level is enough.

But the configuration of the log level is not restricted to that file.  The default is `info`, and if we log in that level in `nxt_http_route.c` it will have important performance issues.  What do you suggest to fix that?  Also, how will you differentiate `info` logs from that file and `info` logs from other files?

> 
> Let the team make the decision of what we should resolve for users, good to discuss.

Reply:
No ideas yet. If the scope is clear, it would be great for reviewers.

Reply:
> So, I'd choose the router process only for now, and leave the possibility of also modifying the main and controller processes in the future (after this release). Does it make sense to you? @lcrilly , @ac000 , does it also make sense to you?

I think so. IIRC the initial impetus for this was to debug decisions the 'router' made.